maven/impl-maven-embedded/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/FileExtractor.java

```diff
@@ -0,0 +1,133 @@
+         extractionIsProcessing.createNewFile();
+         extractionIsProcessing.deleteOnExit();
+
+         return extractionIsProcessing;
+      } catch (IOException e) {
+         throw new RuntimeException(e);
+      }
+   }
+
+   File extract() {
+      File withExtractedDir = checkIfItIsAlreadyExtracted();
+      if (withExtractedDir != null) {
+         return withExtractedDir;
+      }
+      destinationDir.mkdirs();
+      extractFileInDestinationDir();
+      return destinationDir;
+   }
+
+   private void extractFileInDestinationDir() {
+      final File markerFile = createExtractionMarkerFile();
+      final String downloadedPath = fileToExtract.getAbsolutePath();
+
+      try {
+         if (downloadedPath.endsWith(".zip")) {
+            Spacelift.task(fileToExtract, UnzipTool.class).toDir(destinationDir).execute().await();
+         } else if (downloadedPath.endsWith(".tar.gz")) {
+            Spacelift.task(fileToExtract, UntarTool.class).gzip(true).toDir(destinationDir).execute().await();
+         } else if (downloadedPath.endsWith(".tar.bz2")) {
+            Spacelift.task(fileToExtract, UntarTool.class).bzip2(true).toDir(destinationDir).execute().await();
+         } else {
+            throw new IllegalArgumentException(
+               "The distribution " + fileToExtract + " is compressed by unsupported format. "
+                  + "Supported formats are .zip, .tar.gz, .tar.bz2");
+         }
+      } catch (ExecutionException ee) {
+         throw new IllegalStateException(
+            "Something bad happened when the file: " + downloadedPath + " was being extracted. "
+               + "For more information see the stacktrace", ee);
+      }
+      if (!markerFile.delete()) {
+         log.warning("failed to delete temp directory: " + markerFile);
+      }
+      System.out.println(String.format("Resolver: Successfully extracted maven binaries from %s", fileToExtract));
+   }
+
+   private File checkIfItIsAlreadyExtracted() {
+      if (destinationDir.exists() && destinationDir.isDirectory()
+         && destinationDir.list().length > 0 && isExtractionFinished()) {
+         return destinationDir;
+      }
+      return null;
+   }
+
+   private boolean isExtractionFinished() {
+      int count = 0;
+      while (isTempFilePresent() && count < 100) {
+         if (count == 0) {
+            System.out.println(
+               "There is marker file, which means that some other process is already extracting the archive,"
+                  + " waiting for extraction to finish");
+         }
+         System.out.print(".");
+         try {
+            Thread.sleep(100);
+         } catch (InterruptedException e) {
+            log.warning("Problem occurred when the thread was sleeping:\n" + e.getMessage());
+         }
+         count++;
+      }
+      System.out.println();
+      if (count == 99 && isTempFilePresent()) {
+         try {
+            deleteFileRecursively(destinationDir.toPath());
```

Comment:
here it should return false to say that the extraction is not finished and should be started... otherwise, it goes to the line 111 and returns true, which is wrong.

yeah, then the line 111 can just return true without checking if the marker file is present